ts/twitter/src/app/signup/_components/signup-modal.tsx

```diff
@@ -0,0 +1,163 @@
+"use client";
+
+import type { SignupFormValue } from "@/lib/models/signup-form-types";
+import {
+  Box,
+  Button,
+  Flex,
+  FormControl,
+  FormLabel,
+  Input,
+  InputGroup,
+  InputLeftAddon,
+  InputRightElement,
+  Text,
+  VStack,
+} from "@chakra-ui/react";
+import { useRouter } from "next/navigation";
+import { useState } from "react";
+import type React from "react";
+import { FaEye, FaEyeSlash } from "react-icons/fa";
+
+export const SignupModal: React.FC = () => {
+  const [formValues, setFormValues] = useState<SignupFormValue>({
+    displayName: "",
+    username: "",
+    password: "",
+  });
+  const [showPassword, setShowPassword] = useState(false);
+  const router = useRouter();
+
+  // Function to generate a username from display name.
+  const generateUsername = (displayName: string): string => {
+    if (!displayName) return "";
+    return displayName.toLowerCase().replace(/[^\w]/gi, "").substring(0, 15);
+  };
+
+  const handleInputChange = (field: string, value: string) => {
```

Comment:
```suggestion
  const handleInputChange = (field: keyof SignupFormValue, value: string) => {
```
Let's define it as `field: keyof SignupFormValue` to enhance type safety.
This will ensure that only displayName, username, and password fields are accepted, and not any other fields.